From f69a9c7049300ef6b1bef1c874f1946fc53605fd Mon Sep 17 00:00:00 2001
From: Dimitry Ishenko <dimitry.ishenko@gmail.com>
Date: Wed, 12 May 2021 13:03:25 -0400
Subject: [PATCH 02/13] Update m_sampleCount after successfully reading in
 m_Entries

This mostly reverts commit e6c7cda (except for the formatting).
---
 Source/C++/Core/Ap4StszAtom.cpp | 14 ++++++++------
 1 file changed, 8 insertions(+), 6 deletions(-)

diff --git a/Source/C++/Core/Ap4StszAtom.cpp b/Source/C++/Core/Ap4StszAtom.cpp
index e3896e9..225dbbb 100644
--- a/Source/C++/Core/Ap4StszAtom.cpp
+++ b/Source/C++/Core/Ap4StszAtom.cpp
@@ -78,25 +78,27 @@ AP4_StszAtom::AP4_StszAtom(AP4_UI32        size,
     }
 
     stream.ReadUI32(m_SampleSize);
-    stream.ReadUI32(m_SampleCount);
+    AP4_UI32 sampleCount;
+    stream.ReadUI32(sampleCount);
     if (m_SampleSize == 0) { // means that the samples have different sizes
         // check for overflow
-        if (m_SampleCount > (size - AP4_FULL_ATOM_HEADER_SIZE - 8) / 4) {
+        if (sampleCount > (size - AP4_FULL_ATOM_HEADER_SIZE - 8) / 4) {
             return;
         }
         
         // read the entries
-        unsigned char* buffer = new unsigned char[m_SampleCount * 4];
-        AP4_Result result = stream.Read(buffer, m_SampleCount * 4);
+        unsigned char* buffer = new unsigned char[sampleCount * 4];
+        AP4_Result result = stream.Read(buffer, sampleCount * 4);
         if (AP4_FAILED(result)) {
             delete[] buffer;
             return;
         }
-        m_Entries.SetItemCount((AP4_Cardinal)m_SampleCount);
-        for (unsigned int i = 0; i < m_SampleCount; i++) {
+        m_Entries.SetItemCount((AP4_Cardinal)sampleCount);
+        for (unsigned int i = 0; i < sampleCount; i++) {
             m_Entries[i] = AP4_BytesToUInt32BE(&buffer[i * 4]);
         }
         delete[] buffer;
+        m_SampleCount = sampleCount;
     }
 }
 
-- 
2.35.1

